### ISSUE TITLE
Lots of undeclared identifier errors in gcc headers?

### ISSUE THREAD

**Issue by cslux** [Andersbakken/rtags#335]
Hi,

I created a JSON db with ninja for the [chromium!](https://www.chromium.org/Home) code base using the following command.
```
ninja -C build/Debug -t combdb cxx
```
Now when I pass the `compile_commands.json` file to `rdm` executing `rc -J compile_command.json`,
after some time (compile_commands.json is ~80 MB big) it starts parsing the file. However, I get so many errors while processing the source files.
Here is a small snippet of some of the errors, but they are all the same.
```
/usr/lib/gcc/x86_64-redhat-linux/4.9.2/include/emmintrin.h:967:19: error: use of undeclared identifier '__builtin_ia32_punpckhwd128'
/usr/lib/gcc/x86_64-redhat-linux/4.9.2/include/emmintrin.h:973:19: error: use of undeclared identifier '__builtin_ia32_punpckhdq128'
/usr/lib/gcc/x86_64-redhat-linux/4.9.2/include/emmintrin.h:979:19: error: use of undeclared identifier '__builtin_ia32_punpckhqdq128'
/usr/lib/gcc/x86_64-redhat-linux/4.9.2/include/emmintrin.h:985:19: error: use of undeclared identifier '__builtin_ia32_punpcklbw128'
/usr/lib/gcc/x86_64-redhat-linux/4.9.2/include/emmintrin.h:991:19: error: use of undeclared identifier '__builtin_ia32_punpcklwd128'
/usr/lib/gcc/x86_64-redhat-linux/4.9.2/include/emmintrin.h:997:19: error: use of undeclared identifier '__builtin_ia32_punpckldq128'
/usr/lib/gcc/x86_64-redhat-linux/4.9.2/include/emmintrin.h:1003:19: error: use of undeclared identifier '__builtin_ia32_punpcklqdq128'
/usr/lib/gcc/x86_64-redhat-linux/4.9.2/include/emmintrin.h:1009:19: error: use of undeclared identifier '__builtin_ia32_paddb128'
/usr/lib/gcc/x86_64-redhat-linux/4.9.2/include/emmintrin.h:1015:19: error: use of undeclared identifier '__builtin_ia32_paddw128'
/usr/lib/gcc/x86_64-redhat-linux/4.9.2/include/emmintrin.h:1021:19: error: use of undeclared identifier '__builtin_ia32_paddd128'
/usr/lib/gcc/x86_64-redhat-linux/4.9.2/include/emmintrin.h:1027:19: error: use of undeclared identifier '__builtin_ia32_paddq128'
/usr/lib/gcc/x86_64-redhat-linux/4.9.2/include/emmintrin.h:1057:19: error: use of undeclared identifier '__builtin_ia32_psubb128'
/usr/lib/gcc/x86_64-redhat-linux/4.9.2/include/emmintrin.h:1063:19: error: use of undeclared identifier '__builtin_ia32_psubw128'
/usr/lib/gcc/x86_64-redhat-linux/4.9.2/include/emmintrin.h:1069:19: error: use of undeclared identifier '__builtin_ia32_psubd128'
/usr/lib/gcc/x86_64-redhat-linux/4.9.2/include/emmintrin.h:1075:19: error: use of undeclared identifier '__builtin_ia32_psubq128'
/usr/lib/gcc/x86_64-redhat-linux/4.9.2/include/emmintrin.h:1117:19: error: use of undeclared identifier '__builtin_ia32_pmullw128'
/usr/lib/gcc/x86_64-redhat-linux/4.9.2/include/emmintrin.h:1266:19: error: use of undeclared identifier '__builtin_ia32_pand128'
/usr/lib/gcc/x86_64-redhat-linux/4.9.2/include/emmintrin.h:1272:19: error: use of undeclared identifier '__builtin_ia32_pandn128'
/usr/lib/gcc/x86_64-redhat-linux/4.9.2/include/emmintrin.h:1278:19: error: use of undeclared identifier '__builtin_ia32_por128'
/usr/lib/gcc/x86_64-redhat-linux/4.9.2/include/emmintrin.h:1284:19: error: use of undeclared identifier '__builtin_ia32_pxor128'
/usr/lib/gcc/x86_64-redhat-linux/4.9.2/include/emmintrin.h:1290:19: error: use of undeclared identifier '__builtin_ia32_pcmpeqb128'
/usr/lib/gcc/x86_64-redhat-linux/4.9.2/include/emmintrin.h:1296:19: error: use of undeclared identifier '__builtin_ia32_pcmpeqw128'
/usr/lib/gcc/x86_64-redhat-linux/4.9.2/include/emmintrin.h:1302:19: error: use of undeclared identifier '__builtin_ia32_pcmpeqd128'
/usr/lib/gcc/x86_64-redhat-linux/4.9.2/include/emmintrin.h:1308:19: error: use of undeclared identifier '__builtin_ia32_pcmpgtb128'
```

And here is one of the 20000>  compile commands.
```
  {
    "directory": "/home/cschwarzgruber/Developing/chromium/src/.nobackup/out/Debug",
    "command": "../../../third_party/llvm-build/Release+Asserts/bin/clang++ -MMD -MF obj/base/base.async_socket_io_handler_posix.o.d -DV8_DEPRECATION_WARNINGS -D_FILE_OFFSET_BITS=64 -DCHROMIUM_BUILD -DCR_CLANG_REVISION=223108 -DCOMPONENT_BUILD -DUI_COMPOSITOR_IMAGE_TRANSPORT -DUSE_AURA=1 -DUSE_OZONE=1 -DUSE_DEFAULT_RENDER_THEME=1 -DUSE_LIBJPEG_TURBO=1 -DENABLE_ONE_CLICK_SIGNIN -DENABLE_PRE_SYNC_BACKUP -DENABLE_WEBRTC=1 -DENABLE_CONFIGURATION_POLICY -DENABLE_NOTIFICATIONS -DDONT_EMBED_BUILD_METADATA -DENABLE_TASK_MANAGER=1 -DENABLE_EXTENSIONS=1 -DENABLE_SESSION_SERVICE=1 -DENABLE_THEMES=1 -DENABLE_BACKGROUND=1 -DENABLE_GOOGLE_NOW=1 -DCLD_VERSION=2 -DENABLE_SPELLCHECK=1 -DENABLE_CAPTIVE_PORTAL_DETECTION=1 -DENABLE_APP_LIST=1 -DENABLE_SETTINGS_APP=1 -DENABLE_SUPERVISED_USERS=1 -DENABLE_MDNS=1 -DENABLE_SERVICE_DISCOVERY=1 -DV8_USE_EXTERNAL_STARTUP_DATA -DUSE_SYMBOLIZE -DUSE_LIBPCI=1 -DUSE_NSS=1 -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -DBASE_IMPLEMENTATION -DDYNAMIC_ANNOTATIONS_ENABLED=1 -DWTF_USE_DYNAMIC_ANNOTATIONS=1 -D_DEBUG -D_GLIBCXX_DEBUG=1 -Igen -I../../.. -fstack-protector --param=ssp-buffer-size=4 -Werror -pthread -fno-strict-aliasing -Wall -Wno-unused-parameter -Wno-missing-field-initializers -fvisibility=hidden -pipe -fPIC -Wno-reserved-user-defined-literal -Xclang -load -Xclang /home/cschwarzgruber/Developing/chromium/src/third_party/llvm-build/Release+Asserts/lib/libFindBadConstructs.so -Xclang -add-plugin -Xclang find-bad-constructs -Xclang -plugin-arg-find-bad-constructs -Xclang check-weak-ptr-factory-order -Xclang -plugin-arg-find-bad-constructs -Xclang strict-virtual-specifiers -fcolor-diagnostics -B/home/cschwarzgruber/Developing/chromium/src/third_party/binutils/Linux_x64/Release/bin -Wheader-hygiene -Wno-char-subscripts -Wno-unneeded-internal-declaration -Wno-covered-switch-default -Wstring-conversion -Wno-c++11-narrowing -Wno-deprecated-register -Wno-inconsistent-missing-override -Wexit-time-destructors -m64 -march=x86-64 -O0 -g -gdwarf-4 -funwind-tables -gsplit-dwarf -Wno-undefined-bool-conversion -Wno-tautological-undefined-compare -fno-exceptions -fno-rtti -fno-threadsafe-statics -fvisibility-inlines-hidden -Wsign-compare -std=gnu++11  -c ../../../base/async_socket_io_handler_posix.cc -o obj/base/base.async_socket_io_handler_posix.o",
    "file": "../../../base/async_socket_io_handler_posix.cc"
  },
```

Whats interesting about the errors is that they are referring to gcc headers that is pretty weird.
That shouldn't be the case, right?

Thank you for your help!

regards,
  Christian

**Comment by cslux** [comment]
UPDATE:
I just tested it with the rtags project same errors.
```
mkdir build && cd build && cmake ../ -DCMAKE_EXPORT_COMPILE_COMMANDS=true
rdm &
rc -J 
```

**Comment by Andersbakken** [comment]
Hi Christiaehat libclang version is this?

**Comment by cslux** [comment]
Hi,

version is 3.5, installed via the fedora package manager.

BTW it is working for my own cmake based project.

**Comment by Andersbakken** [comment]
Odd. Is your own project c++11-based?

When you start rdm. Does it say something like this:

Includepaths: -isystem /usr/local/lib/clang/3.6.0/include/

Also. Are you passing any other switches to rdm or putting something in
~/.rdmrc

Anders

**Comment by cslux** [comment]
> Is your own project c++11-based?

No c++1y 

> When you start rdm. Does it say something like this:

Yes

```
Includepaths: -isystem /usr/lib64/llvm/clang/3.5.0/include/
```

> Also. Are you passing any other switches to rdm or putting something in ~/.rdmrc

Yes

```
--no-startup-project
--no-rc
```

Christian

**Comment by jhanssen** [comment]
Clang is probably getting confused with the GCC c++1y feature headers, if so then the proper way to solve this is to get a libc++ enabled clang and force a `-stdlib=libc++` argument to all your compile commands. That said if you want to try something quick and dirty then you *might* get away with just downloading libc++ (http://llvm.org/releases/3.5.0/libcxx-3.5.0.src.tar.xz) and pass and `-isystem` pointing to the header directory to rdm.

**Comment by cslux** [comment]
Hi Anders and jhanssen,

I figured it out why clang was so confused. The `-isystem` include path is wrong.

```
Includepaths: -isystem /usr/lib64/llvm/clang/3.5.0/include/
```
should be

```
Includepaths: -isystem /usr/lib/clang/3.5.0/include/
```

But I'm still slightly confused why it had worked the hole time and than suddenly not anymore...hmmm.

Anyway, can I change the default `-isystem` include path to the correct path or do have to pass it to rdm directly? According to the [README](https://github.com/Andersbakken/rtags#finding-clang) it will be assumed that the include files are under *${CLANG_LIBDIR}/clang/CLANG_VERSION_STRING/include*. 

Which in my case would be */usr/lib64/llvm/clang/3.5.0/include/*.
```
llvm-config --libdir
/usr/lib64/llvm
llvm-config --version
3.5.0
```
Is this evaluated at run-time or compile-time? 

Possible "solutions":
* Creating a symlink to /usr/lib64/llvm/clang/3.5.0/include.
* Adding `-isystem /usr/lib/clang/3.5.0/include` to ~/.rdmrc.
* Compiling clang from scratch and setting the install path correctly.
* Adding a cmake define `CLANG_INCDIR`?.
* and some more

What do you think would be the best solution for long term?

Thank you!

Christian

**Comment by Andersbakken** [comment]
Hi Christian. You can override the CLANG_LIBDIR path with a cmake variable
or environment variable. Are you still having issues or did ti all work out?

regards

Anders

**Comment by cslux** [comment]
Hi Anders,

> You can override the CLANG_LIBDIR path with a cmake variable or environment variable.

This does not work as `-lclang` etc. are under /usr/lib64/llvm. However, the include files, for instance `stdinit.h` shipped with clang, are under /usr/lib/clang/3.5.0/include.

Maybe this is a RPM packaging issues. I will try to fetch the package only and modefiy the install path in the ${package}.spec file, so the clang headers will be installed under.
```
${CLANG_LIBDIR}/clang/CLANG_VERSION_STRING/include
```

> Are you still having issues or did ti all work out?

It does work now, if I pass `-isystem /usr/lib/clang/3.5.0/include to `rdm`.
`rdm` output on start-up looks like this.
```
Includepaths: -isystem /usr/lib/clang/3.5.0/include/ -isystem /usr/lib64/llvm/clang/3.5.0/include/
```

regards,
Christian

PS: It works really good, even for the chromium project (Maybe a bit slow when jumping around). However, I shrunk the compile_commands.json from 22000 to 8500 compile commands. But the parsed output still needs 3.5 GB storage (~/.rtags/...). Maybe there are some flags I can pass to `rdm` like `--no-comments` which will help to shrink to DB size.
